pandas/core/util/hashing.py

```diff
@@ -125,7 +129,10 @@ def hash_pandas_object(
                 for _ in [None]
             )
             num_items += 1
-            hashes = itertools.chain(hashes, index_hash_generator)
+
+            # keep `hashes` specifically a generator to keep mypy happy
+            _hashes = itertools.chain(hashes, index_hash_generator)
+            hashes = (x for x in _hashes)
```

Comment:
> you can also do a `cast` here

Our type hints policy is "..., the use of cast is strongly discouraged. Where applicable a refactor of the code to appease static analysis is preferable". https://dev.pandas.io/docs/development/contributing.html#style-guidelines

personally I'm not a fan of changing the runtime behavior of code just to appease mypy if not absolutely necessary unless it makes for cleaner code. (This applies to both our policy and the changes in this PR)

The problem in this function is because Mypy considers the initial assignment as the definition of a variable. https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#type-inference which occurs on L114.
mypy infers the type as `Generator[Any, None, None]`.

it is possible to override the inferred type of a variable by using a variable type annotation https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html#explicit-types-for-variables

This is the approach I took to silence this mypy error in https://github.com/pandas-dev/pandas/pull/28339/commits/304351e23d1586e0cc28108aff9a2a8a9fac5bf7. I prefer this as it has no effect on the runtime behavior.

with regard to the use of a leading underscore to avoid `Incompatible types in assignment..` errors, my preference would be to use a trailing underscore. In PEP8, a trailing underscore is used by convention to avoid conflicts with Python keywords. So I see more similarity here to avoid conflicts than the leading underscore idiom. Not sure what's best here, but just my POV if we want to choose a pattern to be consistent moving forward.

Reply:
It looks like the main actionable suggestion here is to change `_hashes` to `hashes_`; is that right @simonjayhawkins ?  I'm fine with that.